core/src/services/zookeeper/backend.rs

```diff
@@ -0,0 +1,305 @@
+        }
+    }
+
+    fn create_nested_node<'a>(
+        &'a self,
+        path: &'a str,
+        value: &'a [u8],
+    ) -> BoxFuture<'a, Result<()>> {
+        eprintln!("{path}");
+        let mut path = path.to_string();
+        async move {
+            return match self
+                .get_connection()
+                .await?
+                .create(
+                    &path,
+                    value,
+                    &zk::CreateOptions::new(zk::CreateMode::Persistent, self.acl),
+                )
+                .await
+            {
+                Ok(_) => Ok(()),
+                Err(e) => match e {
+                    zk::Error::NoNode => {
+                        let idx = path.rfind('/').unwrap();
+                        let tmpath = path.clone();
+                        path = path.to_string().substring(0, idx).to_string();
+                        if path.as_bytes()[0] != b'/' {
+                            path = "/".to_string() + path.strip_suffix('/').unwrap_or(&path);
+                        }
+                        match self.create_nested_node(&path, value).await {
+                            Ok(()) => {}
+                            Err(e) => {
+                                return Err(Error::new(
+                                    ErrorKind::Unexpected,
+                                    "error from zookeeper",
+                                )
+                                .set_source(e))
+                            }
+                        };
+                        match self
+                            .get_connection()
+                            .await?
+                            .create(
+                                &tmpath,
+                                value,
+                                &zk::CreateOptions::new(zk::CreateMode::Persistent, self.acl),
+                            )
+                            .await
+                        {
+                            Ok(_) => Ok(()),
+                            Err(e) => {
+                                Err(Error::new(ErrorKind::Unexpected, "error from zookeeper")
+                                    .set_source(e))
+                            }
+                        }
+                    }
+                    _ => {
+                        Err(Error::new(ErrorKind::Unexpected, "error from zookeeper").set_source(e))
+                    }
+                },
+            };
+        }
+        .boxed()
+    }
+}
+
+#[async_trait]
+impl kv::Adapter for ZkAdapter {
+    fn metadata(&self) -> kv::Metadata {
+        kv::Metadata::new(
+            Scheme::Zookeeper,
+            "ZooKeeper",
+            Capability {
+                read: true,
+                write: true,
+                ..Default::default()
+            },
+        )
+    }
+
+    async fn get(&self, path: &str) -> Result<Option<Vec<u8>>> {
+        let path = "/".to_string() + path.strip_suffix('/').unwrap_or(path);
```

Comment:
We can use https://opendal.apache.org/docs/rust/opendal/raw/fn.build_rooted_abs_path.html